function.tf

```diff
@@ -0,0 +1,78 @@
+locals {
+  enable_function = var.enable_function || var.enable_extensions
+}
+
+resource "google_service_account" "cloud_functions" {
+  account_id  = "${var.name}-func"
+  description = "A service account for the Observe Cloud Functions"
+}
+
+resource "google_project_iam_member" "cloud_functions" {
+  for_each = toset(var.function_roles)
+
+  project = data.google_project.this.project_id
+  role    = each.key
+  member  = "serviceAccount:${google_service_account.cloud_functions.email}"
+}
+
+resource "google_pubsub_topic_iam_member" "function_pubsub" {
+  topic  = google_pubsub_topic.this.name
+  role   = "roles/pubsub.publisher"
+  member = local.writer_identity
+}
+
+resource "google_cloudfunctions_function" "this" {
+  count = local.enable_function ? 1 : 0
+
+  name                  = var.name
+  description           = "A cloud function that polls data from the GCP REST API"
+  service_account_email = google_service_account.cloud_functions.email
+
+  runtime = "python310"
+  environment_variables = {
+    "PARENT"   = var.resource
+    "TOPIC_ID" = google_pubsub_topic.this.id
+  }
+
+  available_memory_mb = 512
+  trigger_http        = true
+  ingress_settings    = "ALLOW_ALL"
+  timeout             = 120
+
+  source_archive_bucket = var.function_bucket
+  source_archive_object = var.function_object
+  entry_point           = "main"
+
+  labels = var.labels
+}
+
+
+resource "google_service_account" "cloud_scheduler" {
+  account_id  = "${var.name}-sched"
+  description = "A service account to allow the Cloud Scheduler job to trigger a Cloud Function"
+}
+
+resource "google_project_iam_member" "cloud_scheduler_cloud_function_invoker" {
```

Comment:
I would expect a policy binding to specifically allow this service account to invoke a specific function. This is missing that last part (a specific function).

You may need this:

https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function_iam

See also https://cloud.google.com/functions/docs/securing/authenticating#function-to-function